### PR TITLE
Update node 4+, hapi 13+, es6 syntax.  closes #22

### LINKED ISSUE #22: Node >= 4 / es6 updates
Summary of changes required:
- use strict
- `const` and `let` where appropriate (linter will tell you where)
- arrow functions where appropriate (linter will tell you where)
- relevant updates to tests + docs
- updating deps and engine in package.json
- avoid `self` and `+=` / `-=` (not optimized yet for let const)

### PR DESCRIPTION
**Pull request by zoe-1**
closes #22 
#### updated dependencies
- **boom** from 2.x.x. to 3.x.x
- **hoek** from 2.x.x. to 3.x.x
- **hawk** from 3.x.x. to 4.x.x
#### update peerDependencies
- **hapi** from 8.x.x. to 13.x.x
#### update devDependencies
- **code** from 1.x.x to 2.x.x
- **hapi** from 8.x.x to 13.x.x
- **lab** from 6.x.x to 8.x.x
#### updated nodejs engine
- >= 4.0.0 required
#### updated to es6 sytax
- satisfies hapi's linting requirements.
#### 100% test coverage
- shot v3 introduced breaking changes to test/hawk.js.
  -  Previously shot 1.x.x did not process trailers.
    and returned trailers in headers.
  -  See shot/lib/index.js v1 has no trailer processing logic.
    Versus v3 does process trailers. shot's  internals.finish() (./lib/index.js)
    contains logic to process headers and trailers. View this file to see differences
    between shot v3 and v1 header and trailer handling. 
  -  changed tests to verify trailers versus headers in previous version.
- hapi v13 handled bewit errors differently than hapi v8 causing the upgrade to to break bewit tests.
   refactored bewit tests to pass. Project is now ready for [internet falconry ](https://github.com/hueniverse/hawk)
- refactored tests' server init logic.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -28,33 +30,35 @@ internals.hawk = function (server, options) {
...
-                        request.once('finish', function () {
+                        request.once('finish', () => {
 
                             request.plugins['hapi-auth-hawk'] = { payloadHash: Hawk.crypto.finalizePayloadHash(payloadHash) };
                         });
                     });
                 }
 
-                var result = { credentials: credentials, artifacts: artifacts };
+                const result = { credentials: credentials, artifacts: artifacts };
```
Comment (Marsup) [inline lib/index.js]: You should use ES6 shorthands.

File: lib/index.js [hunk 2]
```diff
@@ -62,13 +66,13 @@ internals.hawk = function (server, options) {
                 return reply.continue(result);
             });
         },
-        payload: function (request, reply) {
+        payload: (request, reply) => {
```
Comment (Marsup) [inline lib/index.js]: No need to bind, use simpler form on all keys (`payload(request, reply) {`).

File: lib/index.js [hunk 3]
```diff
@@ -104,20 +112,20 @@ internals.hawk = function (server, options) {
...
     Hoek.assert(options.getCredentialsFunc, 'Missing required getCredentialsFunc method in bewit scheme configuration');
 
-    var settings = Hoek.clone(options);
+    const settings = Hoek.clone(options);
     settings.hawk = settings.hawk || {};
 
-    var scheme = {
-        authenticate: function (request, reply) {
+    const scheme = {
+        authenticate: (request, reply) => {
```
Comment (Marsup) [inline lib/index.js]: Same as up there.

File: lib/index.js [hunk 4]
```diff
@@ -104,20 +112,20 @@ internals.hawk = function (server, options) {
...
 
-            Hawk.server.authenticateBewit(request.raw.req, settings.getCredentialsFunc, settings.hawk, function (err, credentials, bewit) {
+            Hawk.server.authenticateBewit(request.raw.req, settings.getCredentialsFunc, settings.hawk, (err, credentials, bewit) => {
 
-                var result = { credentials: credentials, artifacts: bewit };
+                const result = { credentials: credentials, artifacts: bewit };
```
Comment (Marsup) [inline lib/index.js]: Shorten credentials.

File: package.json [hunk 5]
```diff
@@ -12,23 +12,23 @@
...
-    "code": "1.x.x",
-    "hapi": "8.x.x",
-    "lab": "6.x.x"
+    "code": "2.x.x",
+    "hapi": "13.x.x",
+    "lab": "8.x.x"
   },
   "scripts": {
-    "test": "lab -r console -t 100 -a code -L",
+    "test": "lab -r console -t 100 -a code -L -v",
```
Comment (Marsup) [inline package.json]: Should it stay ?

File: test/hawk.js [hunk 6]
```diff
@@ -52,734 +60,786 @@ describe('hawk scheme', function () {
...
+                };
 
-            Hoek.inherits(TestStream, Stream.Readable);
+                Hoek.inherits(TestStream, Stream.Readable);
 
-            TestStream.prototype._read = function (size) {
+                TestStream.prototype._read = function (size) {
 
-                var self = this;
+                    const self = this;
```
Comment (Marsup) [inline test/hawk.js]: If you're using arrow functions you should drop self.

File: test/hawk.js [hunk 7]
```diff
@@ -52,734 +60,786 @@ describe('hawk scheme', function () {
...
 
-                    done();
+                    getCredentials('john', (err, cred) => {
+
+                        expect(err).to.not.exist();
+
+                        authHeader.artifacts.credentials = cred;
+                        const header = Hawk.server.header(cred, authHeader.artifacts, options);
+
+                        // expect(header).to.equal(res.headers['server-authorization']);
```
Comment (Marsup) [inline test/hawk.js]: Remove that ?

File: test/hawk.js [hunk 8]
```diff
@@ -52,734 +60,786 @@ describe('hawk scheme', function () {
...
 
-                    reply('Success');
-                },
-                config: { auth: 'default' }
-            });
+                expect(err).to.not.exist();
+                server.auth.strategy('default', 'hawk', { getCredentialsFunc: getCredentials });
+                server.route({
+                    method: 'POST', path: '/hawk',
+                    handler: (request, reply) => {
```
Comment (Marsup) [inline test/hawk.js]: More accurate conversion is also `handler(request, reply) {` on all those.

**Comment by zoe-1** [comment]
@Marsup thank you for the review. 
Made the changes you recommended to be made.

File: lib/index.js (hunk 1 above)
Comment (Marsup) [inline lib/index.js]: Missed this one.

**Comment by zoe-1** [comment]
fixed it

**Comment by Marsup** [comment]
That's not what I meant. It should be `const result = { credentials, artifacts };`

**Comment by zoe-1** [comment]
The object needs the keys in it, or it breaks hapi/lib/auth.js logic.
hapi/lib/auth requires the { credentials: xxxxx, artifacts: xxxx } syntax.

**Comment by Marsup** [comment]
http://ariya.ofilabs.com/2013/02/es6-and-object-literal-property-value-shorthand.html §1

**Comment by gergoerdosi** [comment]
`{ credentials, artifacts }` is the same as `{ credentials: credentials, artifacts: artifacts }` in ES6 (object shorthand properties).

**Comment by zoe-1** [comment]
@Marsup thanks you for the sharing the link reference. 
changing it now.  @gergoerdosi above link explains the short hand.

**Comment by zoe-1** [comment]
changed to object shorthand.

**Comment by zoe-1** [comment]
@dannycoates  merged your changes committed today into this PR: 
* removes content-length.
* modified the new test you created to pass hapi 's current linting requirements.

Perhaps you have different plans for this project than what this PR is doing by 
attempting to fulfill #22.  A little feedback would be appreciated. 
cheers!

**Comment by dannycoates** [comment]
@zoe-1 my patch was to fix an immediate issue with the latest releases of node, specifically the HTTP vulnerability. There were users that needed a fix ASAP. Thanks for contributing with this PR and updating the test :)

**Comment by zoe-1** [comment]
@dannycoates thanks for the feedback :)

**Comment by gergoerdosi** [comment]
Any updates on this PR?

**Comment by gergoerdosi** [comment]
Ping @dannycoates.

**Comment by vladikoff** [comment]
ping @dannycoates

**Comment by timoa** [comment]
ping @dannycoates

File: .travis.yml [hunk 9]
```diff
@@ -1,7 +1,8 @@
 language: node_js
 
 node_js:
-  - 0.10
-  - 4
+  - "4.0"
```
Comment (vladikoff) [inline .travis.yml]: do we need 4.0 here or is 4 and 6 good enough?

**Comment by vladikoff** [comment]
I'll take a look at this PR in detail in a bit...

**Comment by zoe-1** [comment]
Closed this PR because it's obsolete.
hapi is now v17 and PR #35 updates the project to v17.

**Pull request closed without merging**